### ISSUE TITLE
Ablation study failure - Training ReactionT5 directly from t5-v1_1-base yields 0% accuracy and repetitive generation

### ISSUE THREAD

**Issue by YangooSen** [sagawatatsuya/ReactionT5v2#7]
Hi authors,

First of all, thank you for sharing your amazing work and code! 

I am currently conducting an ablation study using your code to verify the importance of the first-stage pre-training (CompoundT5). Specifically, I skipped the compound pre-training phase and tried to train the reaction prediction model directly using the raw NLP base model (`google/t5-v1_1-base`) as a starting point.

**What I did:**
1. I set the base model to `google/t5-v1_1-base`.
2. I kept the tokenization process exactly the same (loading the English vocab and adding `additional_tokens.txt`), followed by `model.resize_token_embeddings(len(tokenizer))` to randomly initialize the new chemical tokens.
3. I used `Adafactor` with `learning_rate=1e-4`, `warmup_ratio=0.05`, `bf16=True`, and `max_grad_norm=1.0` to prevent the initial gradient explosion caused by the randomly initialized tokens.
4. I monitored the `eval_loss` instead of `accuracy` for early stopping, since the accuracy stays at 0 for a long time.

Here are my exact bash script and modified training code:

<details>
<summary><b>My Bash Script</b></summary>

```bash
python -u ./model/train.py \
    --is_finetune=0 \
    --model_name_or_path='./t5-v1_1-base' \
    --output_dir='ReactionT5_Without_CompoundT5' \
    --epochs=100 \
    --lr=1e-4 \
    --batch_size=32 \
    --input_max_length=150 \
    --target_max_length=100 \
    --weight_decay=0.01 \
    --evaluation_strategy='epoch' \
    --save_strategy='epoch' \
    --logging_strategy='epoch' \
    --save_total_limit=100 \
    --train_data_path='./datasets/ord/all_ord_reaction_uniq_with_attr20240506_v3_train.csv' \
    --valid_data_path='./datasets/ord/all_ord_reaction_uniq_with_attr20240506_v3_valid.csv' \
    --test_data_path='./datasets/ord/all_ord_reaction_uniq_with_attr20240506_v3_test.csv' \
```
</details>

<details>
<summary><b>My Modified Training Code</b></summary>

```python
import argparse
import os
import sys
import warnings
from pathlib import Path

import datasets
import pandas as pd
import torch
from datasets import Dataset, DatasetDict
from transformers import (
    AutoModelForSeq2SeqLM,
    AutoTokenizer,
    DataCollatorForSeq2Seq,
    EarlyStoppingCallback,
    Seq2SeqTrainer,
    Seq2SeqTrainingArguments,
)

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from utils import (
    add_new_tokens,
    canonicalize,
    filter_out,
    get_accuracy_score,
    preprocess_dataset,
    seed_everything,
    space_clean,
)

# Suppress warnings and disable progress bars
warnings.filterwarnings("ignore")
datasets.utils.logging.disable_progress_bar()


def parse_args():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="Training script for reaction prediction model."
    )
    parser.add_argument(
        "--train_data_path", type=str, required=True, help="Path to training data CSV."
    )
    parser.add_argument(
        "--valid_data_path",
        type=str,
        required=True,
        help="Path to validation data CSV.",
    )
    parser.add_argument("--test_data_path", type=str, help="Path to test data CSV.")
    parser.add_argument(
        "--USPTO_test_data_path",
        type=str,
        help="The path to data used for USPTO testing. CSV file that contains ['REACTANT', 'REAGENT', 'PRODUCT'] columns is expected.",
    )
    parser.add_argument(
        "--output_dir", type=str, default="t5", help="Path of the output directory."
    )
    parser.add_argument(
        "--model_name_or_path",
        type=str,
        required=True,
        help="Model path or name.",
    )
    parser.add_argument(
        "--debug", action="store_true", default=False, help="Enable debug mode."
    )
    parser.add_argument(
        "--epochs",
        type=int,
        default=5,
        help="Number of epochs.",
    )
    parser.add_argument("--lr", type=float, default=1e-3, help="Learning rate.")
    parser.add_argument("--batch_size", type=int, default=16, help="Batch size.")
    parser.add_argument(
        "--input_max_length",
        type=int,
        default=400,
        help="Max input token length.",
    )
    parser.add_argument(
        "--target_max_length",
        type=int,
        default=150,
        help="Max target token length.",
    )
    parser.add_argument(
        "--eval_beams",
        type=int,
        default=5,
        help="Number of beams used for beam search during evaluation.",
    )
    parser.add_argument(
        "--target_column",
        type=str,
        default="PRODUCT",
        help="Target column name.",
    )
    parser.add_argument(
        "--weight_decay",
        type=float,
        default=0.01,
        help="Weight decay.",
    )
    parser.add_argument(
        "--evaluation_strategy",
        type=str,
        default="epoch",
        help="Evaluation strategy used during training. Select from 'no', 'steps', or 'epoch'. If you select 'steps', also give --eval_steps.",
    )
    parser.add_argument(
        "--eval_steps",
        type=int,
        help="Evaluation steps.",
    )
    parser.add_argument(
        "--save_strategy",
        type=str,
        default="epoch",
        help="Save strategy used during training. Select from 'no', 'steps', or 'epoch'. If you select 'steps', also give --save_steps.",
    )
    parser.add_argument(
        "--save_steps",
        type=int,
        default=500,
        help="Save steps.",
    )
    parser.add_argument(
        "--logging_strategy",
        type=str,
        default="epoch",
        help="Logging strategy used during training. Select from 'no', 'steps', or 'epoch'. If you select 'steps', also give --logging_steps.",
    )
    parser.add_argument(
        "--logging_steps",
        type=int,
        default=500,
        help="Logging steps.",
    )
    parser.add_argument(
        "--save_total_limit",
        type=int,
        default=2,
        help="Limit of saved checkpoints.",
    )
    parser.add_argument(
        "--fp16",
        action="store_true",
        default=False,
        help="Enable fp16 training.",
    )
    parser.add_argument(
        "--disable_tqdm",
        action="store_true",
        default=False,
        help="Disable tqdm.",
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=42,
        help="Random seed.",
    )
    parser.add_argument(
        "--is_finetune",
        type=int,
        default=1,
        choices=[0, 1],
        help="1: finetune with compound pretrained weights, 0: train from scratch with random initialization.",
    )

    return parser.parse_args()


def preprocess_metabolism_data(df):
    """
    将代谢数据格式转换为模型输入格式
    """
    # 将 SMILES 列映射到 REACTANT（如果存在）
    if 'SMILES' in df.columns:
        df['REACTANT'] = df['SMILES']
    
    # 其他列填充为空（如果不存在才添加）
    if 'REAGENT' not in df.columns:
        df['REAGENT'] = ' '
    if 'SOLVENT' not in df.columns:
        df['SOLVENT'] = ' '
    if 'PRODUCT' not in df.columns:
        df['PRODUCT'] = 'Metabolites'
    
    return df



def preprocess_df(df, drop_duplicates=True):
    """Preprocess the dataframe by filling NaNs, dropping duplicates, and formatting the input."""
    for col in ["REACTANT", "PRODUCT", "CATALYST", "REAGENT", "SOLVENT"]:
        if col not in df.columns:
            df[col] = None
        df[col] = df[col].fillna(" ")
    if drop_duplicates:
        df = (
            df[["REACTANT", "PRODUCT", "CATALYST", "REAGENT", "SOLVENT"]]
            .drop_duplicates()
            .reset_index(drop=True)
        )
    df["REAGENT"] = df["CATALYST"] + "." + df["REAGENT"] + "." + df["SOLVENT"]
    df["REAGENT"] = df["REAGENT"].apply(lambda x: space_clean(x))
    df["REAGENT"] = df["REAGENT"].apply(lambda x: canonicalize(x) if x != " " else " ")
    df["input"] = "REACTANT:" + df["REACTANT"] + "REAGENT:" + df["REAGENT"]
    return df


def preprocess_USPTO(df):
    df["REACTANT"] = df["REACTANT"].apply(lambda x: str(sorted(x.split("."))))
    df["REAGENT"] = df["REAGENT"].apply(lambda x: str(sorted(x.split("."))))
    df["PRODUCT"] = df["PRODUCT"].apply(lambda x: str(sorted(x.split("."))))

    df["input"] = "REACTANT:" + df["REACTANT"] + "REAGENT:" + df["REAGENT"]
    df["pair"] = df["input"] + " - " + df["PRODUCT"].astype(str)

    return df


if __name__ == "__main__":
    CFG = parse_args()
    CFG.disable_tqdm = True
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    seed_everything(seed=CFG.seed)

    # Load and preprocess data
    train = preprocess_df(
        filter_out(pd.read_csv(CFG.train_data_path), ["REACTANT", "PRODUCT"])
    )
    valid = preprocess_df(
        filter_out(pd.read_csv(CFG.valid_data_path), ["REACTANT", "PRODUCT"])
    )
    if CFG.USPTO_test_data_path:
        train_copy = preprocess_USPTO(train.copy())
        USPTO_test = preprocess_USPTO(pd.read_csv(CFG.USPTO_test_data_path))
        train = train[~train_copy["pair"].isin(USPTO_test["pair"])].reset_index(
            drop=True
        )
    train["pair"] = train["input"] + " - " + train["PRODUCT"]
    valid["pair"] = valid["input"] + " - " + valid["PRODUCT"]
    valid = valid[~valid["pair"].isin(train["pair"])].reset_index(drop=True)
    train.to_csv("train.csv", index=False)
    valid.to_csv("valid.csv", index=False)

    if CFG.test_data_path:
        test = preprocess_df(
            filter_out(pd.read_csv(CFG.test_data_path), ["REACTANT", "PRODUCT"])
        )
        test["pair"] = test["input"] + " - " + test["PRODUCT"]
        test = test[~test["pair"].isin(train["pair"])].reset_index(drop=True)
        test = test.drop_duplicates(subset=["pair"]).reset_index(drop=True)
        test.to_csv("test.csv", index=False)

    dataset = DatasetDict(
        {
            "train": Dataset.from_pandas(train[["input", "PRODUCT"]]),
            "validation": Dataset.from_pandas(valid[["input", "PRODUCT"]]),
        }
    )

    # load tokenizer
    tokenizer = AutoTokenizer.from_pretrained(
        os.path.abspath(CFG.model_name_or_path)
        if os.path.exists(CFG.model_name_or_path)
        else CFG.model_name_or_path,
        return_tensors="pt",
    )
    tokenizer = add_new_tokens(
        tokenizer,
        Path(__file__).resolve().parent.parent / "datasets" / "additional_tokens.txt",
    )
    tokenizer.add_special_tokens(
        {
            "additional_special_tokens": tokenizer.additional_special_tokens
            + ["REACTANT:", "REAGENT:"]
        }
    )
    CFG.tokenizer = tokenizer

    # load model
    if CFG.is_finetune == 1:
        # 微调模式：加载预训练权重
        model = AutoModelForSeq2SeqLM.from_pretrained(
            os.path.abspath(CFG.model_name_or_path)
            if os.path.exists(CFG.model_name_or_path)
            else CFG.model_name_or_path
        ).to(device)
        print(f"✓ Loaded pretrained weights from {CFG.model_name_or_path}")
    else:
        # 从头训练模式：只加载配置，随机初始化权重
        # 消融实验底座：使用普通的 NLP T5 权重作为起点，跳过化学单分子预训练
        nlp_base_model = "t5-v1_1-base"
        model = AutoModelForSeq2SeqLM.from_pretrained(nlp_base_model).to(device)
        print(f"✓ Ablation mode: Loaded NLP weights from {nlp_base_model}, skipping Compound Pre-training")

    model.resize_token_embeddings(len(tokenizer))

    tokenized_datasets = dataset.map(
        lambda examples: preprocess_dataset(examples, CFG),
        batched=True,
        remove_columns=dataset["train"].column_names,
        load_from_cache_file=False,
    )

    data_collator = DataCollatorForSeq2Seq(tokenizer, model=model)

    args = Seq2SeqTrainingArguments(
        CFG.output_dir,
        evaluation_strategy=CFG.evaluation_strategy,
        eval_steps=CFG.eval_steps,
        save_strategy=CFG.save_strategy,
        save_steps=CFG.save_steps,
        logging_strategy=CFG.logging_strategy,
        logging_steps=CFG.logging_steps,
        optim="adafactor",
        warmup_ratio=0.05,
        learning_rate=CFG.lr,
        per_device_train_batch_size=CFG.batch_size,
        per_device_eval_batch_size=CFG.batch_size,
        weight_decay=CFG.weight_decay,
        save_total_limit=CFG.save_total_limit,
        num_train_epochs=CFG.epochs,
        predict_with_generate=True,
        # fp16=CFG.fp16,
        bf16=True,
        disable_tqdm=CFG.disable_tqdm,
        push_to_hub=False,
        load_best_model_at_end=True,
        # metric_for_best_model="accuracy",
        # greater_is_better=True
        metric_for_best_model="loss",
        greater_is_better=False,

        gradient_accumulation_steps=2,     # 如果需要可以改为2或4
        max_grad_norm=1.0,                 # 梯度裁剪
    )

    model.config.eval_beams = CFG.eval_beams
    model.config.max_length = CFG.target_max_length
    trainer = Seq2SeqTrainer(
        model,
        args,
        train_dataset=tokenized_datasets["train"],
        eval_dataset=tokenized_datasets["validation"],
        data_collator=data_collator,
        tokenizer=tokenizer,
        compute_metrics=lambda eval_preds: get_accuracy_score(eval_preds, CFG),
        callbacks=[EarlyStoppingCallback(early_stopping_patience=10)],
    )

    try:
        trainer.train(resume_from_checkpoint=True)
    except:
        trainer.train(resume_from_checkpoint=None)
    # 从头训练：ReactionT5_from_scratch/best_model/
    trainer.save_model(os.path.join(CFG.output_dir, "best_model"))
```
</details>

**What I observed:**
The training loss and eval loss decreased smoothly (e.g., eval loss dropped from 10.22 at Epoch 1 to around 1.54 at Epoch 33). However, the **Exact Match accuracy remained at exactly 0.0**. 

When I manually tested the saved checkpoint (e.g., eval_loss ~ 1.5), I found that the model suffers from severe hallucination and repetitive loops. For example:
* **Input (Reactant + Reagent):** `REACTANT:COC(=O)C1=CCCN(C)C1.O.[Al+3].[H-].[Li+].[Na+].[OH-]REAGENT:C1CCOC1`
* **Ground Truth:** `CN1CCC=C(CO)C1`
* **Model Output:** `CC(C)(C)OC(=O)N1CCN)N1CCN)N1CCN)N1CCN)N1CCN)...` (repeats until `max_length` and has invalid grammar like unclosed parentheses).

**My Questions:**
Since other models like the *Molecular Transformer* can achieve reasonable accuracy by training from scratch on reaction datasets without a massive single-molecule pre-training phase, I am very confused about this result:

1. **Is this complete failure (0% accuracy and repetition) expected** when skipping the CompoundT5 pre-training phase for this specific T5 architecture? 
2. Do you think the pre-existing English NLP knowledge in `t5-v1_1-base` severely interferes with the randomly initialized chemical tokens, causing this mode collapse and repetition? 
3. Did you ever try training from scratch (or from raw T5) during your experiments, and did you observe similar issues?

**Comment by sagawatatsuya** [comment]
Thank you very much for the detailed report, and I apologize for the delayed response.

We have not tested training ReactionT5 directly from `google/t5-v1_1-base`, so I cannot say with certainty whether this behavior is expected.

However, one possibility is that initializing the model from an English-pretrained T5 model may negatively affect training in this setting. In particular, the pre-existing NLP representations may not necessarily be beneficial for SMILES-based reaction modeling and could potentially interfere with learning compared with random initialization. We have not experimentally verified this, so this should be regarded only as a possible explanation.

In our experiments, when training either from random initialization or from CompoundT5 initialization, gradient explosion rarely occurred, and the exact-match accuracy generally increased steadily during training. Therefore, the fact that you observed gradient instability at the beginning and that the accuracy remained at 0% may indicate that the initialization strategy itself is affecting optimization.

Another important point is that several training settings were changed from the original ReactionT5 training configuration. For example, the original forward-prediction training used a learning rate of `1e-3` and the default optimizer in `Seq2SeqTrainingArguments`, whereas your experiment uses Adafactor, `lr=1e-4`, warmup, gradient accumulation, gradient clipping, and bf16 training.

Therefore, to isolate the effect of CompoundT5 pre-training more clearly, I would recommend first keeping the original ReactionT5 training settings unchanged and replacing only the initialization:

1. `CompoundT5 → ORD`
2. `t5-v1_1-base → ORD`
3. `randomly initialized T5 → ORD`

The third setting would also be useful, because loading `t5-v1_1-base` is not equivalent to training from scratch. Comparing raw NLP-pretrained T5 with a randomly initialized T5 may help determine whether the NLP initialization itself is contributing to the training difficulty.

Thank you again for trying this ablation and for reporting the results. I would be interested to know whether the behavior changes when using the original training configuration or random initialization.
